### PR TITLE
Ports - Ngoc & Sav

### PR DESCRIPTION
**Pull request by lebaongoc**
# Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
How did getting up to speed on the existing codebase go? If it went well, what worked? If it didn't go well, what will you do differently next time? | It went well, we looked at the existing codebase and created a diagram showing the inheritance of classes and object compositions. Following that, we answered the comprehension questions and they help guide our understanding of the codebase. In addition, the lecture on References and Values on OO RideShare was very helpful to understand the complex relationships between classes. 
What _inheritance_ relations exist between classes? | Trip, Passenger, and Driver inherited from CsvRecord class
What _composition_ relations exist between classes? | Trip Dispatcher has a one-to-many relationships with Trip, Passenger, and Driver. Trip has a one-to-one relationship with Driver and Passenger. Driver and Passenger have a one-to-many relationship with Trip.
Describe a decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | When we were considering how to handle duration on an in-progress trip, we made a decision to return nil instead of raising an exception so we can better follow the design pattern that we have already laid out. 
Give an example of a _template method_ that you implemented for this assignment | CsvRecord.from_csv
Give an example of a _nominal test_ that you wrote for this assignment | We wrote a test for Driver#total_revenue to check if this method calculates the correct revenue.
Give an example of an _edge case test_ that you wrote for this assignment | We wrote a test for Driver#total_revenue to check if this method calculates the correct revenue when a driver has no trips.
What is a concept that you gained more clarity on as you worked on this assignment | We gained more perspectives on several concepts, including Object Compositions, OOD, TDD (red, green, refactor!), pair programming(ping pong).

### CODE REVIEW

File: lib/trip.rb [hunk 1]
```diff
@@ -1,34 +1,48 @@
-require 'csv'
-
-require_relative 'csv_record'
+require "csv"
+require "awesome_print"
+require "time"
+require_relative "csv_record"
+require_relative "../specs/spec_helper"
```
Comment (kaidamasaki) [inline lib/trip.rb:5]: `spec_helper` shouldn't be required outside of test code.

File: specs/driver_spec.rb [hunk 2]
```diff
@@ -47,55 +49,44 @@
...
-        id: 8,
-        driver: @driver,
-        passenger_id: 3,
-        start_time: "2016-08-08",
-        end_time: "2016-08-08",
-        rating: 5
-      )
-      @driver.add_trip(trip)
-    end
+  before do
```
Comment (kaidamasaki) [inline specs/driver_spec.rb:75]: I think you accidentally moved this out of an inner `describe`.  If you didn't though, you should move your `before` up to the beginning of the `describe`.

File: lib/trip.rb [hunk 3]
```diff
@@ -1,34 +1,48 @@
...
-        raise ArgumentError, 'Passenger or passenger_id is required'
+        raise ArgumentError, "Passenger or passenger_id is required"
       end
 
-      @start_time = start_time
-      @end_time = end_time
+      @start_time = Time.parse(start_time)
+      if end_time
+        @end_time = Time.parse(end_time)
+        if @end_time.to_i < @start_time.to_i
```
Comment (kaidamasaki) [inline lib/trip.rb:28]: You can directly compare `Time` instances.

```suggestion
        if @end_time < @start_time
```

File: lib/driver.rb [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+module RideShare
+  class Driver < CsvRecord
+    attr_reader :name, :vin, :trips
+    attr_accessor :status
+
+    def initialize(id:, name:, vin:, status: :AVAILABLE, trips: nil)
+      super(id)
+      @name = name
+      @vin = vin.length == 17 ? vin : (raise ArgumentError.new("Invalid Vin: #{vin.length} length"))
```
Comment (kaidamasaki) [inline lib/driver.rb:11]: Generally ternaries are only used for things where both branches are values.

I would do:
```suggestion
      if vin.length == 17
        @vin = vin
      else
        raise ArgumentError.new("Invalid Vin: #{vin.length} length")
      end
```
or:
```suggestion
      raise ArgumentError.new("Invalid Vin: #{vin.length} length" unless vin.length == 17
      @vin = vin
```

File: lib/driver.rb [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+    def add_trip(trip)
+      @trips << trip
+    end
+
+    def average_rating(completed_trips = non_inprogress_trips)
+      average_rating = completed_trips.sum do |trip|
+        trip.rating
+      end
+      average_rating /= completed_trips.length.to_f if completed_trips.length > 0
```
Comment (kaidamasaki) [inline lib/driver.rb:28]: I like how you took advantage of the fact that the sum of an empty array is `0` here.  

One small thing: since you return `average_rating` on the very next line I'd probably just directly return it here.
```suggestion
      return average_rating / completed_trips.length.to_f if completed_trips.length > 0
```

File: lib/driver.rb [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+    def non_inprogress_trips
+      return trips.select do |trip|
+               trip.end_time
+             end
+    end
+
+    def total_revenue(completed_trips = non_inprogress_trips)
+      return completed_trips.sum do |trip|
+               (trip.cost - 1.65) * 0.80
```
Comment (kaidamasaki) [inline lib/driver.rb:40]: If the trip cost less than `1.65` the driver will lose money here.  Is that what you intended?

File: lib/trip_dispatcher.rb [hunk 7]
```diff
@@ -19,20 +20,51 @@ def find_passenger(id)
...
-      # Make puts output more useful
+      #Make puts output more useful
       return "#<#{self.class.name}:0x#{object_id.to_s(16)} \
               #{trips.count} trips, \
               #{drivers.count} drivers, \
               #{passengers.count} passengers>"
     end
 
+    def request_trip(passenger_id)
+      assigned_driver = @drivers.find(ifnone = nil) do |driver|
```
Comment (kaidamasaki) [inline lib/trip_dispatcher.rb:37]: Being explicit is fine but `ifnone` defaults to `nil`, so if you don't need to include it if you don't want to.
```suggestion
      assigned_driver = @drivers.find do |driver|
```

File: specs/passenger_spec.rb [hunk 8]
```diff
@@ -34,27 +33,27 @@
...
-        )
-      trip = RideShare::Trip.new(
-        id: 8,
-        passenger: @passenger,
-        start_time: "2016-08-08",
-        end_time: "2016-08-09",
-        rating: 5
-        )
-
+  before do
```
Comment (kaidamasaki) [inline specs/passenger_spec.rb:36]: Same as above.

**Comment by kaidamasaki** [comment]
# Ride Share
## What We're Looking For

Feature|Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | Yes.  I might group commits together a touch more but it looks good! 
Answer comprehension questions | Yes.  One thing to note regarding relations: `Trip` to `Driver` and `Trip` to `Passenger` are many-to-one since many trips can go to one `Driver` or `Passenger`.
**Wave 1** |
Appropriate use of Ruby's `Time` | Yes.
`Trip` has a helper method to calculate duration | Yes.
`Passenger` methods to calculate total cost and total time | Yes.
Complete tests for new behavior | Yes.
**Wave 2** |
`Driver` inherits from `CsvRecord`, and implements `from_csv` | Yes.
`Driver`s and `Trip`s are connected by `TripDispatcher` |  Yes.
`Driver` has method to calculate average rating | Yes.
`Driver` has a method to calculate total revenue | Yes.
**Wave 3** |
`TripDispatcher` has a new method to create trips | Yes.
Trip creation logic is correctly implemented | Yes.
New `Trip` is connected to the appropriate `Driver` and `Passenger` | Yes.
Tests for `request_trip` | Yes.
Wave 1 and 2 methods ignore in-progress trips | Yes.
Tests for Wave 1 and 2 methods w/ in-progress trips | Yes.
**Wave 4 (Optional)** |
`TripDispatcher` assigns trips to the correct driver | N/A
Complete tests | N/A
Appropriate helper methods were made to help with complex logic | N/A
**Overall** |  Great job!  I really liked how much use you made of `Enumerable` methods and how thorough your tests were!
